### PR TITLE
Drop table enhancement

### PR DESCRIPTION
**Pull request by daminichopra**
## Proposed changes

while using Drop table if there is more than one servicepath it will return message ("entity has more than one servicepath",244)
In case no service path is given it will drop whole table with code "204"

## Types of changes

What types of changes does your code introduce to the project: _Put
an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing
      functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after
creating the PR. If you're unsure about any of them, don't hesitate to
ask. We're here to help! This is simply a reminder of what we are going
to look for before merging your code._

- [x] I have read the [CONTRIBUTING][contrib] doc
- [x] I have signed the [CLA][cla]
- [ ] I have added tests that prove my fix is effective or that my
      feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in
      downstream modules

## Further comments

If this is a relatively large or complex change, kick off the discussion
by explaining why you chose the solution you did and what alternatives
you considered, etc...

[cla]: https://fiware.github.io/contribution-requirements/individual-cla.pdf
    "FIWARE Individual Contributor License Agreement"
[contrib]: https://github.com/smartsdk/ngsi-timeseries-api/blob/master/CONTRIBUTING.md
    "Contributing to QuantumLeap"

### CODE REVIEW

File: src/reporter/delete.py
Comment (c0c0n3) [inline src/reporter/delete.py:37]: Shouldn't this be `fiware_sp()`?

Comment (c0c0n3) [reply src/reporter/delete.py:37]: If you feel adventurous, perhaps it's worth renaming those functions to `fiware_service` and `fiware_service_path`, not sure what's going to break though :-)

File: src/reporter/delete.py
Comment (c0c0n3) [inline src/reporter/delete.py:42]: we could also factor this out in a separate function, e.g.

    has_many_service_paths() -> bool:
       if fiware_sp()...

just a suggestion, we can leave it the way it is now if you don't have time for it.

File: src/reporter/delete.py
Comment (c0c0n3) [inline src/reporter/delete.py:43]: I'd change the message to: "request has more than one fiware service path". My thinking is that we don't know here if the table actually contains entities for more than one service path, all we know is what the client put in the request header...

**Comment by c0c0n3** [comment]
@daminichopra thanks for this!! there's only one snag, I think we should call `fiware_sp()` to get the content of the service path header, not `fiware_s()`---see my comments in the review. Also should we add some tests? It looks like the ones we have at the moment aren't enough since they didn't catch we used the fiware service in place of the service path...

**Comment by c0c0n3** [comment]
note to self. we should be able to close #455 after merging this pr.

File: src/reporter/delete.py
Comment (daminichopra) [reply src/reporter/delete.py:37]: Hi @c0c0n3 , In already existing [code](https://github.com/orchestracities/ngsi-timeseries-api/blob/8d6d8e763d9af3dd706da8fb653c820d834ef2e5/src/reporter/delete.py#L36)  fiware_s() was used. Please confirm weather it needs to be checked for only fiware_service_path or both fiware_service or fiware_service-path? Thank you

Comment (c0c0n3) [reply src/reporter/delete.py:37]: Hi @daminichopra, it's line 37 that needs changing I think, I should've linked the comment to that line, sorry about that! Basically I think it should be enough to change line 37 from

    if fiware_s() is not None:

to

    if fiware_sp() is not None:

note the extra `p`! Those functions have horrendous names, no wonder the confusion. I take the blame for that I think it was me who wrote those two `fiware_*` thingies...

**Comment by chicco785** [comment]
@daminichopra any chance you can rebase on top of the master? this way the build should work out, thx

**Comment by daminichopra** [comment]
Hi @chicco785 , Please can u merge this PR if it looks good. Is it necessary to sign CLA to merge this PR?

**Comment by c0c0n3** [comment]
hi @daminichopra it looks like something is broken. I think is a git merge gone sideways, have a look at this

https://github.com/orchestracities/ngsi-timeseries-api/pull/455/files#diff-ac79bcf11021293203657217c0ceed8a7ed24016c4c313322eb4059ddc43d1dfR67

if you can fix it up, rebase on master, then I'll review. thanks sooooo much!!

File: src/reporter/delete.py
Comment (c0c0n3) [inline src/reporter/delete.py:67]: um, isn't this `with` block fishy? it looks like the code as it was before your change? could it be a result of a rebase gone wrong? if you could please double-check what the heck is going on here, but I think we should zap this entire block, i.e. delete lines 55 through 67?

File: src/reporter/delete.py
Comment (c0c0n3) [inline src/reporter/delete.py:69]: after deleting the block above (lines 55 - 67) we should make sure this statement and all the following ones are properly indented, i.e. indentation should start at the same level as the statement

    deleted = trans.delete_entities(....

on line 50?

**Review by c0c0n3 — CHANGES_REQUESTED** [review]

**Comment by c0c0n3** [comment]
@daminichopra there's some still issues w/ this PR, I think it's b/c of a rebase gone wrong, if you could please double-check, see my comments in the review. Also, if you could please:

1. rebase again---there's some new changes in master, but this rebase should be an easy one
2. sign the CLA. you can try checking the box above in the PR description ("I  have signed the CLA") or if that doesn't work, add a new comment with this text: "recheckcla - I have read the CLA Document and I hereby sign the CLA"

thanks alot!!!

File: src/reporter/delete.py
Comment (daminichopra) [reply src/reporter/delete.py]: Hi @c0c0n3 , I have updated changes to PR . Thanks

Comment (c0c0n3) [reply src/reporter/delete.py]: @daminichopra that's great, thx soooo much!

**Review by c0c0n3 — APPROVED** [review]

**Comment by c0c0n3** [comment]
@daminichopra thx for the new commits, I've approved the PR. If you could pls do one last thing for me: sing the CLA. To do that

1. try checking the box above in the PR description ("I have signed the CLA") or if that doesn't work
2. add a new comment to this PR with this text: "recheckcla - I have read the CLA Document and I hereby sign the CLA"

thanks alot!!!

**Comment by daminichopra** [comment]
> @daminichopra thx for the new commits, I've approved the PR. If you could pls do one last thing for me: sing the CLA. To do that
> 
> 1. try checking the box above in the PR description ("I have signed the CLA") or if that doesn't work
> 2. add a new comment to this PR with this text: "recheckcla - I have read the CLA Document and I hereby sign the CLA"
> 
> thanks alot!!!

Hi @c0c0n3 , Now confirming the signing process inside the company, so please wait for a while.

**Comment by daminichopra** [comment]
Hi @c0c0n3 @chicco785 ,  I have signed the CLA for this PR.

**Comment by daminichopra** [comment]
I have read the CLA Document and I hereby sign the CLA.

File: src/reporter/delete.py
Comment (chicco785) [inline src/reporter/delete.py:37]: @c0c0n3 rather than the list of service paths in the request, should we drop the table if and only if

request header include a single service path and that service path is the only service path available in the table?

the rational is that with the current code logic, it seems to me you can drop a whole table despite it includes more than one service path.

**Comment by chicco785** [comment]
@daminichopra let's see what @c0c0n3 think about my comment, if i got it wrong, we are ready to merge!

**Comment by c0c0n3** [comment]
@chicco785 @daminichopra yes good points. Looking back at #406 what you say makes sense, so ideally the implementation should be

1. if request has no service path ==> drop table
2. if request has multiple service paths ==> 422 error
3. if request has one service path p and p is the only service path in the table ==> drop table
4. else ==> 422 error

At the moment we can't do (3) unless we extend the translator w/ an operation to check if condition (3) is true. @daminichopra do you have enough dev cycles for this additional piece of work? If not, we could just return an error if the  request has any service path in it, just to be on the safe side. So in this case the implementation could be:

1. if request has no service path ==> drop table
2. if request has one or many service paths ==> 422 error

In this scenario we see `dropTable` just as a maintenance convenience knob you can turn if you don't want to write yourself the SQL to drop the table and (2) would be some kind of safety net whereby if there's service paths, we assume the caller didn't mean to specify the `dropTable` param but rather call the bare delete entities method which already deletes all the records in the table having the specified service paths. Not optimal though.

Thoughts?

**Comment by daminichopra** [comment]
> @chicco785 @daminichopra yes good points. Looking back at #406 what you say makes sense, so ideally the implementation should be
> 
> 1. if request has no service path ==> drop table
> 2. if request has multiple service paths ==> 422 error
> 3. if request has one service path p and p is the only service path in the table ==> drop table
> 4. else ==> 422 error
> 
> At the moment we can't do (3) unless we extend the translator w/ an operation to check if condition (3) is true. @daminichopra do you have enough dev cycles for this additional piece of work? If not, we could just return an error if the request has any service path in it, just to be on the safe side. So in this case the implementation could be:
> 
> 1. if request has no service path ==> drop table
> 2. if request has one or many service paths ==> 422 error
> 
> In this scenario we see `dropTable` just as a maintenance convenience knob you can turn if you don't want to write yourself the SQL to drop the table and (2) would be some kind of safety net whereby if there's service paths, we assume the caller didn't mean to specify the `dropTable` param but rather call the bare delete entities method which already deletes all the records in the table having the specified service paths. Not optimal though.
> 
> Thoughts?

At present the suggested code fix incorporated below  two comments:
1. if request has no service path ==> drop table
2. if request has one or many service paths ==> 422 error

For point (3) "if request has one service path p and p is the only service path in the table ==> drop table".
I would suggest to raise a separate issue for  point (3) which can be fixed separately as an enhancement. Thank you :)

**Comment by chicco785** [comment]
> At present the suggested code fix incorporated below two comments:
> 
> 1. if request has no service path ==> drop table
> 2. if request has one or many service paths ==> 422 error
> 
> For point (3) "if request has one service path p and p is the only service path in the table ==> drop table".
> I would suggest to raise a separate issue for point (3) which can be fixed separately as an enhancement. Thank you :)

agree with the path proposed

**Comment by daminichopra** [comment]
Hi @c0c0n3 , Please close this PR if it looks good.

**Review by chicco785 — COMMENTED** [review]
it seems you have overwritten `docs/manuals/user/index.md` with the main `docs/manuals/index.md`, any reason for doing so?

**Comment by daminichopra** [comment]
> it seems you have overwritten `docs/manuals/user/index.md` with the main `docs/manuals/index.md`, any reason for doing so?
@chicco785 , I have updated fix for this [Issue](https://github.com/orchestracities/ngsi-timeseries-api/issues/345)

**Comment by chicco785** [comment]
> > it seems you have overwritten `docs/manuals/user/index.md` with the main `docs/manuals/index.md`, any reason for doing so?
> > @chicco785 , I have updated fix for this [Issue](https://github.com/orchestracities/ngsi-timeseries-api/issues/345)

i see, but the issue was not the content per se, but how readthedocs generates the nav bar, i fixed this in a separate pr #503 , can you remove changes to this file, so we can proceed?

thx

**Comment by daminichopra** [comment]
> > > it seems you have overwritten `docs/manuals/user/index.md` with the main `docs/manuals/index.md`, any reason for doing so?
> > > @chicco785 , I have updated fix for this [Issue](https://github.com/orchestracities/ngsi-timeseries-api/issues/345)
> 
> i see, but the issue was not the content per se, but how readthedocs generates the nav bar, i fixed this in a separate pr #503 , can you remove changes to this file, so we can proceed?
> 
> thx

Removed Changes and updated commit.

**Comment by daminichopra** [comment]
@c0c0n3 @chicco785 , Gentle Reminder!! :)

**Comment by chicco785** [comment]
> @c0c0n3 @chicco785 , Gentle Reminder!! :)

1) build is failing, can't be merged.
2) I believe it does not implement what @c0c0n3 suggested:

> At present the suggested code fix incorporated below two comments:
> 
> if request has no service path ==> drop table
> if request has one or many service paths ==> 422 error

I feel the logic is more:
if request has many service paths ==> 422 error
if request has one service path => drop table

if this is the case, it's not what we want.

**Comment by chicco785** [comment]
@daminichopra i cleaned the pr, and fixed the logic.

as by previous comment, it's better to proceed as follow:
1. define test cases
2. develop code supporting test cases

**Comment by daminichopra** [comment]
> @daminichopra i cleaned the pr, and fixed the logic.
> 
> as by previous comment, it's better to proceed as follow:
> 
> 1. define test cases
> 2. develop code supporting test cases

@chicco785 , Thanks for this i got it. Please let me know if further something needs to be done in this PR.

**Comment by daminichopra** [comment]
> > @daminichopra i cleaned the pr, and fixed the logic.
> > as by previous comment, it's better to proceed as follow:
> > 
> > 1. define test cases
> > 2. develop code supporting test cases
> 
> @chicco785 , Thanks for this i got it. Please let me know if further modification required for the closure of this PR.

Soft Reminder!!

**Review by chicco785 — APPROVED** [review]

**Pull request merged**
